From 3dd0dfc152b0ee92e6ef00ecb5a82989ece76f34 Mon Sep 17 00:00:00 2001
From: Priyadarshnee Srinivasan <priys@qti.qualcomm.com>
Date: Fri, 27 Jun 2025 14:47:06 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL wifi:ath12k:Add module params for retail
 configuration

Fetch the deployment type advertised by HALPHY during service ready and
use it to set indoor/outdoor service flags in the WMI_INIT command.
Introduce module parameters to read user configuration for AFC support
and reg_no_action. Advertise these settings to the firmware and upper
layers via wiphy extended features.
Add two new ext features: retail AFC support and general AFC support.

Patch-dependency:
0346-02-wifi-ath12k-Add-support-for-processing-AFC-power-eve.patch
0346-04-wifi-ath12k-Send-power-update-complete-event-after-A.patch
0346-06-wifi-ath12k-Extract-AFC-response-and-send-AFC-cmd-to.patch
0048-cfg80211-mac80211-regulatory-awgn-and-6g-channel-cha.patch

Feature: AFC
Tag: 6G_REGULATORY
Signed-off-by: Priyadarshnee Srinivasan <priys@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/core.h |  4 ++
 drivers/net/wireless/ath/ath12k/mac.c  | 14 +++++
 drivers/net/wireless/ath/ath12k/reg.c  | 34 ++++++++++--
 drivers/net/wireless/ath/ath12k/reg.h  |  3 +-
 drivers/net/wireless/ath/ath12k/wmi.c  | 71 ++++++++++++++++++++------
 drivers/net/wireless/ath/ath12k/wmi.h  | 23 +++++++--
 include/uapi/linux/nl80211.h           |  9 ++++
 7 files changed, 132 insertions(+), 26 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 2bbf989..e653a79 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -1420,6 +1420,10 @@ struct ath12k_base {
 	 * response
 	 */
 	struct ath12k_6ghz_sp_reg_rule *sp_rule;
+	/* afc_dev_deployment is used to store the device deployment type
+	 * as advertised by FW.
+	 */
+	enum ath12k_afc_dev_deploy_type afc_dev_deployment;
 
 	/* Current DFS Regulatory */
 	enum ath12k_dfs_region dfs_region;
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 11b83b1..d605261 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -18865,6 +18865,20 @@ static int ath12k_mac_hw_register(struct ath12k_hw *ah)
 				      NL80211_EXT_FEATURE_FILS_DISCOVERY);
 		wiphy_ext_feature_set(wiphy,
 				      NL80211_EXT_FEATURE_UNSOL_BCAST_PROBE_RESP);
+		/* For 6 GHz radios, check if both FW and Host support AFC feature and if so,
+		 * advertise the AFC feature support to the higher layers.
+		 */
+		if (test_bit(WMI_TLV_SERVICE_AFC_SUPPORT, ar->ab->wmi_ab.svc_map) &&
+		    ath12k_6ghz_sp_pwrmode_supp_enabled) {
+		    wiphy_ext_feature_set(hw->wiphy, NL80211_EXT_FEATURE_TARGET_AND_HOST_AFC_SUPPORT);
+		    /* If reg_no_action module param is set, user wishes to operate in
+		     * enterprise mode of AFC.
+		     */
+		    if (!ath12k_afc_reg_no_action) {
+			wiphy_ext_feature_set(hw->wiphy, NL80211_EXT_FEATURE_RETAIL_AFC_SUPPORT);
+			ath12k_dbg(ar->ab, ATH12K_DBG_MAC, "Sending retail AFC feature support to higher layers\n");
+		    }
+		}
 	}
 
 	wiphy_ext_feature_set(wiphy, NL80211_EXT_FEATURE_PUNCT);
diff --git a/drivers/net/wireless/ath/ath12k/reg.c b/drivers/net/wireless/ath/ath12k/reg.c
index 42e4de9..e8313a8 100644
--- a/drivers/net/wireless/ath/ath12k/reg.c
+++ b/drivers/net/wireless/ath/ath12k/reg.c
@@ -21,10 +21,34 @@
 #define ETSI_WEATHER_RADAR_BAND_HIGH		5650
 #define ETSI_WEATHER_RADAR_BAND_CAC_TIMEOUT	600000
 
-/* These variables need to assigned to true for enabling Proxy AFC, else 0 */
-bool ath12k_afc_disable_timer_check = true;
-bool ath12k_afc_disable_req_id_check = true;
-bool ath12k_afc_test_enabled = true;
+/* The default config is enabling retail AFC */
+/* When ath12k_afc_disable_timer_check = false, the AP will enable all
+ * the AFC timers in HALPHY.
+ */
+bool ath12k_afc_disable_timer_check;
+module_param_named(afc_disable_timer_check, ath12k_afc_disable_timer_check, bool, 0644);
+MODULE_PARM_DESC(afc_disable_timer_check, "Disable AFC expiry timer check for enterprise mode: 1-disable, 0-enable");
+
+/* When ath12k_afc_disable_req_id_check = false, the AP will enable request
+ * ID match in the FW.
+ */
+bool ath12k_afc_disable_req_id_check;
+module_param_named(afc_disable_req_id_check, ath12k_afc_disable_req_id_check, bool, 0644);
+MODULE_PARM_DESC(afc_disable_req_id_check, "Disable AFC request id check for enterprise mode: 1-disable, 0-enable");
+
+/* Turn off AFC support in the host. The host driver will notify the firmware
+ * that AFC is not supported,
+ */
+bool ath12k_6ghz_sp_pwrmode_supp_enabled;
+module_param_named(6ghz_sp_pwrmode_supp_enabled, ath12k_6ghz_sp_pwrmode_supp_enabled, bool, 0644);
+MODULE_PARM_DESC(6ghz_sp_pwrmode_supp_enabled, "Enable AFC support to FW: 1-enable, 0-disable");
+
+/* When ath12k_afc_reg_no_action = false, the AP will trigger channel
+ * change on AFC response.
+ */
+bool ath12k_afc_reg_no_action;
+module_param_named(afc_reg_no_action, ath12k_afc_reg_no_action, bool, 0644);
+MODULE_PARM_DESC(afc_reg_no_action, "Do not trigger channel change on AFC response: 1-enable, 0-disable");
 
 /* Given a global opclass number create the corresponding  array token.
  * Examples:
@@ -2210,7 +2234,7 @@ static struct ath12k_afc_location *ath12k_reg_fill_afc_location_obj(struct ath12
 	if (!p_afc_location)
 		return NULL;
 
-	p_afc_location->deployment_type = 0;
+	p_afc_location->deployment_type = ar->ab->afc_dev_deployment;
 
 	return p_afc_location;
 }
diff --git a/drivers/net/wireless/ath/ath12k/reg.h b/drivers/net/wireless/ath/ath12k/reg.h
index dae6f51..f174ffc 100644
--- a/drivers/net/wireless/ath/ath12k/reg.h
+++ b/drivers/net/wireless/ath/ath12k/reg.h
@@ -29,7 +29,8 @@ struct ath12k;
 
 extern bool ath12k_afc_disable_timer_check;
 extern bool ath12k_afc_disable_req_id_check;
-extern bool ath12k_afc_test_enabled;
+extern bool ath12k_afc_reg_no_action;
+extern bool ath12k_6ghz_sp_pwrmode_supp_enabled;
 
 /* DFS regdomains supported by Firmware */
 enum ath12k_dfs_region {
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index f6320ce..1c1c141 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -75,6 +75,10 @@ struct ath12k_wmi_svc_rdy_ext_parse {
 	bool dma_ring_cap_done;
 };
 
+/**
+ * ath12k_wmi_svc_rdy_ext2_arg - WMI service ready extended 2
+ * @afc_deployment_type: AFC deployment type (indoor, outdoor)
+ */
 struct ath12k_wmi_svc_rdy_ext2_arg {
 	u32 reg_db_version;
 	u32 hw_min_max_tx_power_2ghz;
@@ -88,6 +92,7 @@ struct ath12k_wmi_svc_rdy_ext2_arg {
 	u32 max_num_linkview_peers;
 	u32 max_num_msduq_supported_per_tid;
 	u32 default_num_msduq_supported_per_tid;
+	u32 afc_deployment_type;
 };
 
 struct ath12k_wmi_svc_rdy_ext2_parse {
@@ -5263,6 +5268,10 @@ ath12k_fill_band_to_mac_param(struct ath12k_base  *soc,
 	}
 }
 
+#define cond_set_or_clear_flag(cond, flags, bit) \
+	((cond) ? set_bit((bit), (volatile unsigned long *)&(flags)) : \
+	 clear_bit((bit), (volatile unsigned long *)(&(flags))))
+
 static void
 ath12k_wmi_copy_resource_config(struct ath12k_base *ab,
 				struct ath12k_wmi_resource_config_params *wmi_cfg,
@@ -5331,20 +5340,26 @@ ath12k_wmi_copy_resource_config(struct ath12k_base *ab,
 	wmi_cfg->flags2 = le32_encode_bits(tg_cfg->peer_metadata_ver,
 					   WMI_RSRC_CFG_FLAGS2_RX_PEER_METADATA_VERSION);
 
-	if (tg_cfg->afc_support) {
-		wmi_cfg->host_service_flags &= ~(1 << WMI_RSRC_CFG_HOST_SUPPORT_LP_SP_MODE_BIT);
-		wmi_cfg->host_service_flags |= 1 << WMI_RSRC_CFG_HOST_SUPPORT_LP_SP_MODE_BIT;
-	}
-
-	if (tg_cfg->afc_disable_timer_check) {
-		wmi_cfg->host_service_flags &= ~(1 << WMI_RSRC_CFG_HOST_AFC_DIS_TIMER_CHECK_BIT);
-		wmi_cfg->host_service_flags |= 1 << WMI_RSRC_CFG_HOST_AFC_DIS_TIMER_CHECK_BIT;
-	}
-
-	if (tg_cfg->afc_disable_req_id_check) {
-		wmi_cfg->host_service_flags &= ~(1 << WMI_RSRC_CFG_HOST_AFC_DIS_REQ_ID_CHECK_BIT);
-		wmi_cfg->host_service_flags |= 1 << WMI_RSRC_CFG_HOST_AFC_DIS_REQ_ID_CHECK_BIT;
-	}
+	cond_set_or_clear_flag(tg_cfg->afc_support, wmi_cfg->host_service_flags,
+			       WMI_RSRC_CFG_HOST_SUPPORT_LP_SP_MODE_BIT);
+	cond_set_or_clear_flag(tg_cfg->afc_disable_timer_check,
+			       wmi_cfg->host_service_flags,
+			       WMI_RSRC_CFG_HOST_AFC_DIS_TIMER_CHECK_BIT);
+	cond_set_or_clear_flag(tg_cfg->afc_disable_req_id_check,
+			       wmi_cfg->host_service_flags,
+			       WMI_RSRC_CFG_HOST_AFC_DIS_REQ_ID_CHECK_BIT);
+	cond_set_or_clear_flag(tg_cfg->afc_indoor_support,
+			       wmi_cfg->host_service_flags,
+			       WMI_RSRC_CFG_HOST_AFC_INDOOR_SUPPORT);
+	cond_set_or_clear_flag(tg_cfg->afc_outdoor_support,
+			       wmi_cfg->host_service_flags,
+			       WMI_RSRC_CFG_HOST_AFC_OUTDOOR_SUPPORT);
+
+	ath12k_dbg(NULL, ATH12K_DBG_WMI, "afc_support: %u "
+		   "dis_timer_check: %u, dis_reqid_check: %u, indoor: %u, outdoor: %u\n",
+		   tg_cfg->afc_support, tg_cfg->afc_disable_timer_check,
+		   tg_cfg->afc_disable_req_id_check, tg_cfg->afc_indoor_support,
+		   tg_cfg->afc_outdoor_support);
 
 	wmi_cfg->host_service_flags = cpu_to_le32(tg_cfg->is_reg_cc_ext_event_supported <<
 				WMI_RSRC_CFG_HOST_SVC_FLAG_REG_CC_EXT_SUPPORT_BIT);
@@ -5369,9 +5384,27 @@ ath12k_wmi_copy_resource_config(struct ath12k_base *ab,
 					    WMI_RSRC_CFG_EMA_INIT_CONFIG_BEACON_SIZE));
 }
 
-void ath12k_set_afc_config(struct ath12k_wmi_resource_config_arg *config)
+/**
+ * ath12k_set_afc_config() - Set AFC config parameters
+ * @config: WMI resource config
+ * @ab: ath12k base
+ *
+ * Return: None
+ */
+static void ath12k_set_afc_config(struct ath12k_wmi_resource_config_arg *config,
+				  struct ath12k_base *ab)
 {
-	config->afc_support = ath12k_afc_test_enabled;
+	config->afc_indoor_support = false;
+	config->afc_outdoor_support = false;
+
+	if (ab->afc_dev_deployment == ATH12K_AFC_DEPLOYMENT_INDOOR)
+		config->afc_indoor_support = true;
+	else if (ab->afc_dev_deployment == ATH12K_AFC_DEPLOYMENT_OUTDOOR)
+		config->afc_outdoor_support = true;
+	else if (ab->afc_dev_deployment == ATH12K_AFC_DEPLOYMENT_UNKNOWN)
+		config->afc_indoor_support = true;
+
+	config->afc_support = ath12k_6ghz_sp_pwrmode_supp_enabled;
 	config->afc_disable_timer_check = ath12k_afc_disable_timer_check;
 	config->afc_disable_req_id_check = ath12k_afc_disable_req_id_check;
 }
@@ -5597,7 +5630,9 @@ int ath12k_wmi_cmd_init(struct ath12k_base *ab)
 	if (test_bit(WMI_SERVICE_WDS_NULL_FRAME_SUPPORT, ab->wmi_ab.svc_map))
 		arg.res_cfg.is_wds_null_frame_supported = true;
 
-	ath12k_set_afc_config(&arg.res_cfg);
+	if (test_bit(WMI_TLV_SERVICE_AFC_SUPPORT, ab->wmi_ab.svc_map))
+		ath12k_set_afc_config(&arg.res_cfg, ab);
+
 	ab->hw_params->wmi_init(ab, &arg.res_cfg);
 	ab->wow.wmi_conf_rx_decap_mode = arg.res_cfg.rx_decap_mode;
 
@@ -6347,6 +6382,7 @@ static int ath12k_pull_svc_ready_ext2(struct ath12k_wmi_pdev *wmi_handle,
 	arg->max_user_per_ppdu_ofdma = le32_to_cpu(ev->max_user_per_ppdu_ofdma);
 	arg->max_user_per_ppdu_mumimo = le32_to_cpu(ev->max_user_per_ppdu_mumimo);
 	arg->target_cap_flags = le32_to_cpu(ev->target_cap_flags);
+	arg->afc_deployment_type = le32_to_cpu(ev->afc_deployment_type);
 	return 0;
 }
 
@@ -6510,6 +6546,7 @@ static int ath12k_wmi_svc_rdy_ext2_parse(struct ath12k_base *ab,
 			return ret;
 		}
 		ab->chwidth_num_peer_caps = parse->arg.chwidth_num_peer_caps;
+		ab->afc_dev_deployment = parse->arg.afc_deployment_type;
 		break;
 
 	case WMI_TAG_ARRAY_STRUCT:
diff --git a/drivers/net/wireless/ath/ath12k/wmi.h b/drivers/net/wireless/ath/ath12k/wmi.h
index a80458d..efd96bc 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -2605,6 +2605,14 @@ enum ath12k_peer_metadata_version {
 	ATH12K_PEER_METADATA_V1B
 };
 
+/**
+ * struct ath12k_wmi_resource_config_arg - resource configuration to FW
+ * @afc_support: Bool to indicate if host/user supports AFC
+ * @afc_disable_timer_check: Bool to indicate if user wishes to disable AFC timer check in FW
+ * @afc_disable_req_id_check: Bool to indicate if user wishes to disable AFC req id check in FW
+ * @afc_indoor_support: Bool to indicate if indoor deployment is supported
+ * @afc_outdoor_support: Bool to indicate if outdoor deployment is supported
+ */
 struct ath12k_wmi_resource_config_arg {
 	u32 num_vdevs;
 	u32 num_peers;
@@ -2677,6 +2685,8 @@ struct ath12k_wmi_resource_config_arg {
 	u32 afc_support;
 	u32 afc_disable_timer_check;
 	u32 afc_disable_req_id_check;
+	bool afc_indoor_support;
+	bool afc_outdoor_support;
 };
 
 struct ath12k_wmi_init_cmd_arg {
@@ -2773,9 +2783,11 @@ struct wmi_ctrl_path_pmlo_telemetry_stats {
 #define WMI_RSRC_CFG_FLAG1_THREE_WAY_COEX_CONFIG_OVERRIDE_SUPPORT BIT(25)
 #define WMI_PDEV_MEC_AGING_TIMER_THRESHOLD_VALUE 5000
 #define WMI_RSRC_CFG_EMA_INIT_CONFIG_BEACON_SIZE               GENMASK(15, 0)
-#define WMI_RSRC_CFG_HOST_SUPPORT_LP_SP_MODE_BIT              7
-#define WMI_RSRC_CFG_HOST_AFC_DIS_TIMER_CHECK_BIT             8
-#define WMI_RSRC_CFG_HOST_AFC_DIS_REQ_ID_CHECK_BIT            9
+#define WMI_RSRC_CFG_HOST_SUPPORT_LP_SP_MODE_BIT               7
+#define WMI_RSRC_CFG_HOST_AFC_DIS_TIMER_CHECK_BIT              8
+#define WMI_RSRC_CFG_HOST_AFC_DIS_REQ_ID_CHECK_BIT             9
+#define WMI_RSRC_CFG_HOST_AFC_INDOOR_SUPPORT                  10
+#define WMI_RSRC_CFG_HOST_AFC_OUTDOOR_SUPPORT                 11
 
 struct ath12k_wmi_resource_config_params {
 	__le32 tlv_header;
@@ -3032,6 +3044,10 @@ enum wmi_channel_width {
 #define WMI_EHT_MCS_NSS_10_11  GENMASK(11, 8)
 #define WMI_EHT_MCS_NSS_12_13  GENMASK(15, 12)
 
+/**
+ * struct wmi_service_ready_ext2_event - extended service ready event params
+ * @afc_deployment_type: AFC deployment type indicated by FW (indoor,outdoor)
+ */
 struct wmi_service_ready_ext2_event {
 	__le32 reg_db_version;
 	__le32 hw_min_max_tx_power_2ghz;
@@ -3045,6 +3061,7 @@ struct wmi_service_ready_ext2_event {
 	__le32 max_num_linkview_peers;
 	__le32 max_num_msduq_supported_per_tid;
 	__le32 default_num_msduq_supported_per_tid;
+	__le32 afc_deployment_type;
 } __packed;
 
 struct ath12k_wmi_caps_ext_params {
diff --git a/include/uapi/linux/nl80211.h b/include/uapi/linux/nl80211.h
index 8100342..b0c23c1 100644
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -6898,6 +6898,13 @@ enum nl80211_feature_flags {
  *
  * @NL80211_EXT_FEATURE_ERP: Driver supports ErP low power mode.
  *
+ * @NL80211_EXT_FEATURE_TARGET_AND_HOST_AFC_SUPPORT: Both the host and the
+ *       target supports AFC (Automatic Frequency Control) feature.
+ *
+ * @NL80211_EXT_FEATURE_RETAIL_AFC_SUPPORT: The device supports retail AFC
+ *       mode where channel selection after AFC response will be done
+ *       by the application layers.
+ *
  * @NUM_NL80211_EXT_FEATURES: number of extended features.
  * @MAX_NL80211_EXT_FEATURES: highest extended feature index.
  */
@@ -6979,6 +6986,8 @@ enum nl80211_ext_feature_index {
 	NL80211_EXT_FEATURE_DEVICE_BW,
 	NL80211_EXT_FEATURE_MLD_LINK_REMOVAL_OFFLOAD,
 	NL80211_EXT_FEATURE_ERP,
+	NL80211_EXT_FEATURE_TARGET_AND_HOST_AFC_SUPPORT,
+	NL80211_EXT_FEATURE_RETAIL_AFC_SUPPORT,
 
 	/* add new features before the definition below */
 	NUM_NL80211_EXT_FEATURES,
-- 
2.34.1

